### PR TITLE
feat: proof validation implemented in experimental-api

### LINKED ISSUE #619: feature(experimental): add proof validation
https://github.com/eigerco/beerus/blob/96aceb2140c56361136c439d1f8acd269a2e0ce7/crates/experimental-api/src/rpc.rs#L293

### PR DESCRIPTION
**Pull request by ICavlek**
Proof validation exists in `crates::core::storage_proofs`. Due to new requirements, it has been also implemented for the new experimental-api. Branch workflow is as follows:

- Creation of unit tests in the` crates::core::storage_proofs` to check the original implementation
- Migration of implementation to `crates::core::experimental_api::gen`
- Migration of unit tests to validate that it has the same functionality
- Code cleanup

Things to be checked with more details:

- Unit tests that have been created - does something need to be added or changed
- `jsonrpc::error` - code value for the errors is it correctly done (possible feature `impl From<FromStrError>` to kick out boilerplate `map_err()?` or different implementation with the current scope)
- usage of `starknet::FieldElemen`t due to pedersen and poseidon hash - inside `parse_proof()` conversion from `gen::Felt` to `starknet::FieldElement` so that pedersen and poseidon hash functions could be called (additional usage of `felt_to_bits()` and `felt_from_bits()` from `crates::core::utils`)
- Beerus Client added to experimental-api bin - necessary to extract L1 block_num and state_root from node 

(Closes #619)

### CODE REVIEW

File: crates/experimental-api/src/gen.rs [hunk 1]
```diff
@@ -102,7 +112,7 @@ pub mod gen {
 
     #[derive(Clone, Debug, Deserialize, Serialize)]
     #[serde(try_from = "i64")]
-    pub struct BlockNumber(i64);
+    pub struct BlockNumber(pub i64);
```
Comment (sergey-melnychuk) [inline crates/experimental-api/src/gen.rs:115]: Now `BlockNumer` can be negative :)

File: crates/experimental-api/src/gen.rs [hunk 2]
```diff
@@ -2162,6 +2185,312 @@ pub mod gen {
         pub state_commitment: Option<Felt>,
     }
 
+    impl GetProofResult {
```
Comment (sergey-melnychuk) [inline crates/experimental-api/src/gen.rs:2188]: Please move out the implementation out of `gen.rs`, it is only for DTOs and relevant validation.

File: crates/experimental-api/src/gen.rs [hunk 3]
```diff
@@ -2137,7 +2160,7 @@ pub mod gen {
         pub transaction_hash: Option<Felt>,
     }
 
-    #[derive(Clone, Debug, Deserialize, Serialize)]
+    #[derive(Clone, Debug, Deserialize, Serialize, Default)]
```
Comment (sergey-melnychuk) [inline crates/experimental-api/src/gen.rs:2163]: What exactly does `Default` for `ContractData` mean? What such a contract can do?

File: crates/experimental-api/src/gen.rs [hunk 4]
```diff
@@ -1659,6 +1676,12 @@ pub mod gen {
                 &self.0
             }
         }
+
+        impl From<StorageKey> for String {
+            fn from(value: StorageKey) -> Self {
+                value.0
```
Comment (sergey-melnychuk) [inline crates/experimental-api/src/gen.rs:1682]: There is already `impl AsRef` doing exactly this

File: crates/experimental-api/src/gen.rs [hunk 5]
```diff
@@ -996,6 +1007,12 @@ pub mod gen {
                 &self.0
             }
         }
+
+        impl From<Address> for Felt {
+            fn from(value: Address) -> Self {
+                value.0
```
Comment (sergey-melnychuk) [inline crates/experimental-api/src/gen.rs:1013]: There is already `AsRef`

File: crates/experimental-api/src/gen.rs [hunk 6]
```diff
@@ -954,13 +964,14 @@ pub mod gen {
         pub unit: PriceUnit,
     }
 
-    #[derive(Clone, Debug, Deserialize, Serialize)]
+    #[derive(Clone, Debug, Deserialize, Serialize, Default, PartialEq)]
```
Comment (sergey-melnychuk) [inline crates/experimental-api/src/gen.rs:967]: What is a "default Felt"? Is it zero? Empty sting? It is unclear, thus a user needs to explicitly state what they need.

File: crates/experimental-api/src/gen.rs [hunk 7]
```diff
@@ -9,10 +9,20 @@ pub mod gen {
...
+
+    use beerus_core::storage_proofs::types::Direction;
+    use beerus_core::utils::{felt_from_bits, felt_to_bits};
 
     #[derive(Clone, Debug, Deserialize, Serialize)]
     pub struct Address(pub Felt);
 
+    impl From<Address> for String {
+        fn from(value: Address) -> String {
+            value.0 .0
```
Comment (sergey-melnychuk) [inline crates/experimental-api/src/gen.rs:22]: `.as_ref().as_ref()` would work instead

File: crates/experimental-api/src/gen.rs [hunk 8]
```diff
@@ -9,10 +9,20 @@ pub mod gen {
     use serde_json::Value;
 
     use iamgroot::jsonrpc;
+    use starknet_crypto::{pedersen_hash, poseidon_hash_many, FieldElement};
```
Comment (sergey-melnychuk) [inline crates/experimental-api/src/gen.rs:12]: In general `gen.rs` contains only generated code and is not expected to be updater directly.

If something is missing, it is added to the codegen tool, so on each change of the spec painful merges can be avoided.

I will put this into README/CONTRIBUTING.

File: crates/experimental-api/src/rpc.rs [hunk 9]
```diff
@@ -291,6 +299,28 @@ impl gen::Rpc for Context {
         key: StorageKey,
         block_id: BlockId,
     ) -> std::result::Result<Felt, jsonrpc::Error> {
+        let l1_block_num = self.node.read().await.l1_block_number as i64;
```
Comment (sergey-melnychuk) [inline crates/experimental-api/src/rpc.rs:302]: This block seems like a good candidate for a method/function extraction (and maybe relevant unit-testing if it makes sense).

**Review by sergey-melnychuk — CHANGES_REQUESTED** [review]

File: crates/experimental-api/src/gen.rs (hunk 1 above)
Comment (ICavlek) [reply crates/experimental-api/src/gen.rs]: Changed back to private

File: crates/experimental-api/src/gen.rs (hunk 2 above)
Comment (ICavlek) [reply crates/experimental-api/src/gen.rs]: Migrated to newly created proof_validation module

File: crates/experimental-api/src/gen.rs (hunk 3 above)
Comment (ICavlek) [reply crates/experimental-api/src/gen.rs]: Removed Default

File: crates/experimental-api/src/gen.rs (hunk 4 above)
Comment (ICavlek) [reply crates/experimental-api/src/gen.rs]: Removed

File: crates/experimental-api/src/gen.rs (hunk 5 above)
Comment (ICavlek) [reply crates/experimental-api/src/gen.rs]: Removed

File: crates/experimental-api/src/gen.rs (hunk 6 above)
Comment (ICavlek) [reply crates/experimental-api/src/gen.rs]: Removed Default

File: crates/experimental-api/src/gen.rs (hunk 7 above)
Comment (ICavlek) [reply crates/experimental-api/src/gen.rs]: Removed

File: crates/experimental-api/src/gen.rs (hunk 8 above)
Comment (ICavlek) [reply crates/experimental-api/src/gen.rs]: Removed

File: crates/experimental-api/src/rpc.rs (hunk 9 above)
Comment (ICavlek) [reply crates/experimental-api/src/rpc.rs]: Migrated to `impl Context`. Added proper unit tests

File: crates/experimental-api/tests/rpc.rs [hunk 10]
```diff
@@ -252,8 +252,7 @@ async fn test_getStorageAt() -> Result<(), Error> {
         "0x0341c1bdfd89f69748aa00b5742b03adbffd79b8e80cab5c50d91cd8c2a79be1",
     )?;
 
-    let block_id =
-        BlockId::BlockNumber { block_number: BlockNumber::try_new(600612)? };
```
Comment (sergey-melnychuk) [inline crates/experimental-api/tests/rpc.rs:256]: why this change is necessary?

File: crates/experimental-api/src/rpc.rs [hunk 11]
```diff
@@ -103,6 +110,40 @@ impl IntoResponse for RpcError {
...
+                let mut block_num = l1_block_num;
+                let user_request_block_num = *block_number.as_ref();
+                if user_request_block_num < block_num {
+                    block_num = user_request_block_num;
+                }
+                Ok(BlockId::BlockNumber {
+                    block_number: BlockNumber::try_new(block_num)?,
+                })
+            }
+            gen::BlockId::BlockHash { block_hash } => {
```
Comment (sergey-melnychuk) [inline crates/experimental-api/src/rpc.rs:133]: For block hashes it might be trickier that that. Ideally we should resolve the block hash into the block number and check if that block number is acceptable.

Please create a follow up issue for that and put a comment here referencing the issue, just to avoid scope extension for this pull request.

Comment (ICavlek) [reply crates/experimental-api/src/rpc.rs:135]: Created https://github.com/eigerco/beerus/issues/675 and added to backlog of the [beerus](https://github.com/orgs/eigerco/projects/27) github project

File: crates/experimental-api/tests/rpc.rs (hunk 10 above)
Comment (ICavlek) [reply crates/experimental-api/tests/rpc.rs]: I have pushed the last commit with the revert back to 600612 for visualization. There is also a change in `ctx()` setup of node's `l1_state_root`. 

In `crates::experimental-api::src::proof_validation::mod.rs` there is a check `*state_commitment == parsed_global_root && global_root == parsed_global_root`. 

They didn't match because of the wrong `l1_state_root`, making the test fail and I have adapted `block_id` at first. Both tests are valid, with `Tag::Latest` and 600612, but when I was testing it on my local pc, I was constantly using `Tag::Latest` for calling rpc and it was easier for me to update the test with these values.
